src/signHash.c

```diff
@@ -0,0 +1,290 @@
+// screen are displayed. This function is passed to UX_DISPLAY, which calls it
+// on each element of the screen. It should return NULL for elements that
+// should not be displayed, and otherwise return the element itself. Elements
+// can be identified by their userid.
+//
+// For the comparison screen, we use the preprocessor to make the scroll
+// buttons more intuitive: we only display them if there is more text hidden
+// off-screen.
+//
+// Note that we did not define a preprocessor for the approval screen. This is
+// because we always want to display every element of that screen. The
+// preprocessor acts a filter that selectively hides elements; since we did
+// not want to hide any elements, no preprocessor was necessary.
+static const bagl_element_t* ui_prepro_signHash_compare(const bagl_element_t *element) {
+	switch (element->component.userid) {
+	case 1:
+		// 0x01 is the left icon (see screen definition above), so return NULL
+		// if we're displaying the beginning of the text.
+		return (ctx->displayIndex == 0) ? NULL : element;
+	case 2:
+		// 0x02 is the right, so return NULL if we're displaying the end of the text.
+		return (ctx->displayIndex == sizeof(ctx->hexHash)-12) ? NULL : element;
+	default:
+		// Always display all other elements.
+		return element;
+	}
+}
+
+// This is the button handler for the comparison screen. Unlike the approval
+// button handler, this handler doesn't send any data to the computer.
+static unsigned int ui_signHash_compare_button(unsigned int button_mask, unsigned int button_mask_counter) {
+	switch (button_mask) {
+	// The available button mask values are LEFT, RIGHT, EVT_RELEASED, and
+	// EVT_FAST. EVT_FAST is set when a button is held for 8 "ticks," i.e.
+	// 800ms.
+	//
+	// The comparison screens in the Zilliqa app allow the user to scroll using
+	// the left and right buttons. The user should be able to hold a button
+	// and scroll at a constant rate. When the user first presses the left
+	// button, we'll hit the LEFT case; after they've held the button for 8
+	// ticks, we'll hit the EVT_FAST | LEFT case. Since we want to scroll at a
+	// constant rate regardless, we handle both cases identically.
+	//
+	// Also note that, unlike the approval screen, we don't check for
+	// EVT_RELEASED. In fact, when a single button is released, none of the
+	// switch cases will be hit, so we'll stop scrolling.
+	case BUTTON_LEFT:
+	case BUTTON_EVT_FAST | BUTTON_LEFT: // SEEK LEFT
+		// Decrement the displayIndex when the left button is pressed (or held).
+		if (ctx->displayIndex > 0) {
+			ctx->displayIndex--;
+		}
+		// Use the displayIndex to recalculate the displayed portion of the
+		// text. os_memmove is the Ledger SDK's version of memmove (there is
+		// no os_memcpy). In practice, I don't think it matters whether you
+		// use os_memmove or the standard memmove from <string.h>.
+		os_memmove(ctx->partialHashStr, ctx->hexHash+ctx->displayIndex, 12);
```

Comment:
There is indeed no difference. Feel free to use memcpy / memmove from string.h, as you already include it for strcpy etc.